File: 1261. Find Elements in a Contaminated Binary Tree/_1261.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class FindElements:
    def __init__(self, root: TreeNode):
        if root:
            root.val = 0
            self.recover(root)
        self.root = root

    def recover(self, root: TreeNode):
        if root.left:
            root.left.val = root.val*2+1
            if root.left.left or root.left.right:
                self.recover(root.left)
        if root.right:
            root.right.val = root.val*2+2
            if root.right.left or root.right.right:
                self.recover(root.right)

    def find(self, target: int) -> bool:
        res = list()
        res.append(target)
        while target!=0:
            if target % 2 == 1:
                target = (target-1)/2
                res.append(target)
            else:
                target = (target-2)/2
                res.append(target)
        res.reverse()
        root = self.root
        for i in range(len(res)):
            if not root:
                return False
            if root.val != res[i]:
                return False
            if i == len(res)-1:
                return True
            if res[i+1]%2 == 0:
                root = root.right
            else:
                root = root.left

class FindElements1:
    def __init__(self, root: TreeNode):
        self.seen = set()
        
        def dfs(node: TreeNode, v: int) -> None:
            if node:
                node.val = v    
                self.seen.add(v)
                dfs(node.left, 2 * v + 1)
                dfs(node.right, 2 * v + 2)
            
        dfs(root, 0)
        
    def find(self, target: int) -> bool:
        return target in self.seen


# Your FindElements object will be instantiated and called as such:
# obj = FindElements(root)
# param_1 = obj.find(target)

File: 1261. Find Elements in a Contaminated Binary Tree/1261.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
package leetcode

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

type FindElements struct {
    root *TreeNode
}


func Constructor(root *TreeNode) FindElements {
    if root != nil{
		root.Val = 0
	}
	recover(root)
	var newroot  = FindElements{root}
	return newroot
}

func recover(root *TreeNode){
	if root.Left != nil{
		root.Left.Val = root.Val*2+1
		recover(root.Left)
	}
	if root.Right != nil{
		root.Right.Val = root.Val*2+2
		recover(root.Right)
	}
}


func (this *FindElements) Find(target int) bool {
	res := make([]int,0)
	res = append(res, target)
	for target != 0{
		if target%2 == 0{
			target = (target-2)/2
		}else{
			target = (target-1)/2
		}
		res = append(res, target)
	}

	tmp := this.root
	for i:=len(res)-1;i>=0;i--{
		if tmp == nil{
			return false
		}
		if tmp.Val != res[i]{
			return false
		}
		if i==0{
			return true
		}
		if res[i-1]%2==0{
			tmp = tmp.Right
		}else{
			tmp = tmp.Left
		}
	}
	return true
}


/**
 * Your FindElements object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Find(target);
 */